devito/petsc/solve.py

```diff
@@ -27,13 +28,21 @@ def PETScSolve(eq, target, bcs=None, solver_parameters=None, **kwargs):
                    halo=target.halo[1:] if is_time_dep else target.halo)
         for prefix in ['y_matvec', 'x_matvec', 'b_tmp']]
 
-    # TODO: Extend to rearrange equation for implicit time stepping.
-    matvecaction = MatVecEq(y_matvec, LinearSolveExpr(eq.lhs.subs(target, x_matvec),
-                            target=target, solver_parameters=solver_parameters),
-                            subdomain=eq.subdomain)
+    b, F_target = separate_eqn(eq, target)
+
+    # Args were updated so need to update target to enable uxreplace on F_target
+    new_target = {func for func in retrieve_functions(F_target) if
+                  func.function == target.function}.pop()
```

Comment:
May be safer to do this in two steps? i.e.
```
new_target = {f for f in retrieve_functions(F_target) if f.function == target.function}
assert len(new_target) == 1  # Sanity check: one target expected
new_target = new_target.pop()
```
That way if you ever end up with multiple targets for whatever reason, it will error, rather than failing cryptically